### ISSUE TITLE
Testing: compatibility with -no-migrations mode

### ISSUE THREAD

**Issue by benjaminrigaud** [BezBartek/django-db-views#24]
When running `pytest --no-migrations`, Models are synced to their corresponding tables as-is: no migrations are executed.

`django-db-views` doesn't support this mode: the views are not created and the tests fail.

As a workaround, I have implemented a fixture to be able to apply any migration on the fly.

A proper implementation in `django-db-views` could be to uncouple the SQL creating the views from the migration framework: it would permit the creation of a few helper to execute this SQL outside of the migrations (`django-pgtrigger` does something like that).

What are your thoughts?

References:
- https://pytest-django.readthedocs.io/en/latest/database.html#no-migrations-disable-django-migrations
- https://django-pgtrigger.readthedocs.io/en/4.6.0/faq.html#how-do-i-test-triggers

**Comment by BezBartek** [comment]
@benjaminrigaud 
Thank you for your thoughts
I will try to get deeper into it through the weekend

**Comment by benjaminrigaud** [comment]
Hi,

If I implement a fix, would you have time to review it? :bow: 

Ben

**Comment by BezBartek** [comment]
Hello @benjaminrigaud 
Please do so, thank you!
Apologize that I didn't find time to write it
I will review your code and merge

**Comment by lajus** [comment]
Hello,

I have run into the same issue while experimenting with django-db-views on my project. 

@benjaminrigaud you said
> As a workaround, I have implemented a fixture to be able to apply any migration on the fly.

would it be possible for you to share the code of your fixture ?

Thanks,
Jonathan

**Comment by benjaminrigaud** [comment]
> would it be possible for you to share the code of your fixture ?

Sorry @lajus, I can't share that code from a private Repo.

Some details:
- It uses `__import__` to import the migration module
- Then instantiates the `Migration` class from this module
- Then it calls `migration.apply(...)`

**Comment by stefanofusai** [comment]
Also interested in this

**Comment by BezBartek** [comment]
@stefanofusai  @lajus 
I almost finish, it will take additional week cuz huge loads at my work currently

**Comment by stefanofusai** [comment]
Thank you so much! Let me know if I can be of help

**Comment by BezBartek** [comment]
@stefanofusai 
check fixture that I added, it does work.

**Comment by stefanofusai** [comment]
@BezBartek thank you so much for your work but unfortunately it's not working for me :(

I'm getting errors when I remove the `--no-migrations`, likely due to incompatibility with [factory-boy](https://github.com/FactoryBoy/factory_boy/)
Doing this: `MyModelFactory.create()` results in the following error:
```
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <django.db.backends.postgresql.base.Cursor [closed] [IDLE] (host=... user=... database=...) at 0x7ff292c2b290>
query = 'INSERT INTO "myapp_mymodel" (...) VALUES (...) RETURNING "myapp_mymodel"."id"'
params = (...)

    def execute(
        self,
        query: Query,
        params: Params | None = None,
        *,
        prepare: bool | None = None,
        binary: bool | None = None,
    ) -> Self:
        """
        Execute a query or command to the database.
        """
        try:
            with self._conn.lock:
                self._conn.wait(
                    self._execute_gen(query, params, prepare=prepare, binary=binary)
                )
        except e._NO_TRACEBACK as ex:
>           raise ex.with_traceback(None)
E           django.db.utils.ProgrammingError: non è possibile modificare la vista materializzata "myapp_mymodel"
```

The error roughly translates to, in English: `cannot modify the materialized view "myapp_mymodel"`.
Here is more code that may be relevant:
```py
class MyModel(DBMaterializedView): # my materialized view model
    ...

    view_definition = ...
    
    class Meta:
        db_table = ...
        managed = False

class MyModelFactory(DjangoModelFactory): # my factory-boy factory
    class Meta:
        django_get_or_create = ...
        model = MyModel

    ...

# In my global conftest, I have to put this otherwise for some reason the models aren't created correctly in the database (even if running with --no-migrations)
# Not sure if this separate issue can also be looked into
def pytest_sessionstart() -> None: 
    MyModelFactory._meta.managed = True  # noqa: SLF001

```

**Comment by BezBartek** [comment]
@stefanofusai 
You can't use factories to create `MyModel` instances, because django factories does inserts into DB, and you can't insert into View by definition.

**Comment by stefanofusai** [comment]
@BezBartek you're right :)
I solved this by deleting `MyModelFactory` and doing `MyModel.refresh()` in my `django_db_setup` fixture, after all other models have been created by factories. Let me know if this is the correct approach!

**Comment by stefanofusai** [comment]
@BezBartek hello, I tried using your fixture to enable --no-migrations but I'm getting the following errors:
```
 +-+---------------- 1 ----------------
    | Traceback (most recent call last):
    |   File "REDACTED/.venv/lib64/python3.13/site-packages/_pytest/runner.py", line 546, in teardown_exact
    |     fin()
    |     ~~~^^
    |   File "REDACTED/.venv/lib64/python3.13/site-packages/_pytest/fixtures.py", line 1032, in finish
    |     raise exceptions[0]
    |   File "REDACTED/.venv/lib64/python3.13/site-packages/_pytest/fixtures.py", line 1021, in finish
    |     fin()
    |     ~~~^^
    |   File "REDACTED/.venv/lib64/python3.13/site-packages/django_db_views/fixtures.py", line 46, in no_migrations_teardown
    |     ViewMigrationAutoDetector.get_backward_migration_class(view_model)(
    |     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^
    |   File "REDACTED/.venv/lib64/python3.13/site-packages/django_db_views/autodetector.py", line 244, in get_backward_migration_class
    |     if issubclass(model, DBMaterializedView):
    |        ~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^
    | TypeError: issubclass() arg 1 must be a class
```

Do you have an example of how your fixture should be used?

**Comment by BezBartek** [comment]
I just import it to conftext.py, you don't need to use it anywhere explicitly as fixture, it has autouse.

"Also, pytest doesn't find the fixture unless I import it explicitly (from django_db_views.fixtures import django_db_views_setup). Is that normal?" - yes it is

**Comment by stefanofusai** [comment]
> [<img alt="" width="16" height="16" src="https://avatars.githubusercontent.com/u/46556949?u=430ef5c145193417ec95cbf819292536ba97d3ae&amp;v=4&amp;size=80">@stefanofusai](https://github.com/stefanofusai?rgh-link-date=2025-01-22T20%3A02%3A46.000Z) I just import it to conftext.py, you don't need to use it anywhere explicitly as fixture, it has autouse.
> 
> "Also, pytest doesn't find the fixture unless I import it explicitly (from django_db_views.fixtures import django_db_views_setup). Is that normal?" - yes it is

Thank you for the reply. Unfortunately it stil doesn't work as it gives the following error at **tests teardown**: `TypeError: issubclass() arg 1 must be a class`

```
_________________________________________ ERROR at teardown of TestWatchListings.test_list[active-request_data0-v4] __________________________________________
[gw6] darwin -- Python 3.13.0 

    def no_migrations_teardown() -> None:
        view_models = ViewMigrationAutoDetector.get_current_view_models()
        with django_db_blocker.unblock(), connection.schema_editor() as schema_editor:
            engine = schema_editor.connection.settings_dict["ENGINE"]
            for view_model in view_models:
                backward_migration = (
>                   ViewMigrationAutoDetector.get_backward_migration_class(view_model)(
                        "",
                        view_model._meta.db_table,
                        engine=engine,
                    )
                )

.venv/lib/python3.13/site-packages/django_db_views/fixtures.py:46: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

model = ('customers', 'watchdata')

    @staticmethod
    def get_backward_migration_class(model) -> Type[BackwardViewMigrationBase]:
>       if issubclass(model, DBMaterializedView):
E       TypeError: issubclass() arg 1 must be a class

.venv/lib/python3.13/site-packages/django_db_views/autodetector.py:244: TypeError
------------------------------------------------------------------ Captured stderr teardown ------------------------------------------------------------------
```

**Comment by stefanofusai** [comment]
@BezBartek fixed by #42
